API template - overloads api page cleanup

sjtrimble opened the pull request:
- API page clean up primarily focusing on new overloads design
- Moved some SCSS files to corresponding sections (some were in the modules section and should be in the page-specific sections)

**Functionality Todos:**
- Add relevant index of overload
- Make individual overload collapsable
- Show only the first overload expanded, rest collapsed
- Text should change to _'collapse all'_ once _'show all'_ is selected
- Need to fix chevron/carrot rotation animation when overloads / overload item is expanded or collapsed
- Need to display `object ...` as the and type , and the first sentence of the description should be "_This is an anonymous type with a **signature** (link) described above._".

<hr>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```
## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #24349

## What is the new behavior?

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

sjtrimble committed:
build(docs-infra): improve API overload templates

mary-poppins commented:
You can preview 2bc5c65 at https://pr24976-2bc5c65.ngbuilds.io/.

mary-poppins commented:
You can preview 8dfcc78 at https://pr24976-8dfcc78.ngbuilds.io/.

petebacondarwin commented:
I'm afraid this needs a significant rebase.
Will the additional todos be completed in another PR?
Visually the "overloads" H4 looks smaller and less significant than the H5 headings of each overload, which seems wrong.
What else has changed, it is not so easy to tell from the diff. Perhaps some before and after images of significant changes would be helpful to speed up the approval process?

sjtrimble commented:
@petebacondarwin 
Here are some of the changes visualized (before/after):

**Methods Section Changes**
![methods-comparison](https://user-images.githubusercontent.com/2958442/43100377-079a6786-8e7a-11e8-918e-7b9748323b1b.png)
- Edit/source code icons blue to match link styling
- Smaller code font to match rest of text size
- Removal of box/border around parameters table

**Overloads Section Changes**
![overload-initial](https://user-images.githubusercontent.com/2958442/43100380-0a15c208-8e7a-11e8-9cf6-5ceb66549071.png)
- Edit/source code icons blue to match link styling
- Larger heading size for Overloads section heading
- Removal of box/border around overloads section heading
- Expand/collapse caret icon accompanied by text for more clear instructions (as individual overload item section will have their own collapse/expand sections in the new design)

**Overload Details Changes**
![overload-detail](https://user-images.githubusercontent.com/2958442/43100379-09e9f5a6-8e7a-11e8-991f-0e5e52b5c572.png)
- Added overload item heading with # of # to easily better distinguish between individual overload sections
- Smaller code font to match rest of text size
- Removal of box/border around parameters table
- Updated styling of individual overload section sub-headings so content hierarchy is a bit more clear

sjtrimble commented:
@petebacondarwin 

1. Regarding the additional to-do's, I was hoping you or someone on the team could help :)
2. Overloads heading is an error and I will fix
3. I will try to rebase and see how it goes.

sjtrimble commented:
Design for reference:
![api-page-overloads](https://user-images.githubusercontent.com/2958442/43100902-a3bdceb8-8e7b-11e8-9a91-2751b06e2246.png)

sjtrimble commented:
Overloads parameter table design discussed Jul 26 (_for reference_):

![overloads-param-table-design](https://user-images.githubusercontent.com/2958442/43288887-220d69ea-90de-11e8-96c0-cfdc17fb90f4.png)

IgorMinar commented:
indent + no color looks good to me.

the option without indentation looks better on this mock, but once it's in the context of the entire page the lack of indentation makes the visual hierarchy harder to see, so that's why I think we should try to stick with the indentation for now.

mary-poppins commented:
You can preview a8f5cd5 at https://pr24976-a8f5cd5.ngbuilds.io/.

mary-poppins commented:
You can preview 07bd5c7 at https://pr24976-07bd5c7.ngbuilds.io/.

sjtrimble commented:
@petebacondarwin 

I rebased, and made the changes based on the mockup we agreed on.
There's a list of _Functionality Todos_ in the description above that covers all of the items that are missing from this PR that I'll need help with :)

petebacondarwin committed:
build(docs-infra): add method overload index

petebacondarwin committed:
build(docs-infra): fix individual API overload templates

* Make individual overloads collapsible
* Show only the first overload expanded, rest collapsed
* Text changes to 'collapse all' once 'show all' is clicked
* Fix chevron/carrot rotation animation when overloads / overload item is expanded or collapsed

petebacondarwin committed:
build(docs-infra): display long overload parameter types as `object`

In some overloads, the parameter type can be a large anonymous
object type.
This change displays such types as `object`. It is then up to the
documentation author to put more information about the type in the
method usage notes.

petebacondarwin commented:
I have rebased and force-pushed with a few more commits that fix the outstanding requirements.

petebacondarwin commented:
I created a PR https://github.com/angular/angular/pull/25448 that shows the change to `HttpClient.request` in action. I excluded that commit from this PR to help with PullApprove.

mary-poppins commented:
You can preview e79c9ce at https://pr24976-e79c9ce.ngbuilds.io/.

IgorMinar commented:
I like the formatting of the "params" and "returns" sections. I however think that collapsing all overloads makes the page much harder to navigate. I think these overloads are verbose and that's what is is, trying to hide that via collapsed section under an opaque heading "Overload X of Y" makes the navigation only more confusing. I suggest that we get rid of the ability to collapse individual overloads and leave it at that.

jenniferfell commented:
@IgorMinar and I just met to review this design. Definitely an improvement over what is there today. It's much easier to see where each overload begins and ends. 

Requests: 
* Overloads section title: We like the current title "17 overloads..." more than the more general "Overloads" in this PR. 
* Overload titles: "Overload 1 of 17" can simply be "Overload 1". The "of 17" doesn't add anything important. 
* Signature: We prefer regular font over bold. Regular matches what is used for signatures at top of API pages, so lets keep that standard. 
* Show v. collapse: We like the "Show all" and "Collapse all" toggle. We like the default with all collapsed. We'd like to simplify what happens with "Show all." We'd like show all to fully expand all the overloads, rather than just expand to a list of the overloads. Also, Igor doesn't think that individual expand-collapse is that useful, so we can just remove that ability. I agree that the two-tier expand-collapse causes some interaction confusion for what "Show all" and "Collapse all" will do, so removing the secondary expand-collapse helps. Also, search only works on what's expanded, so "Show all" expanding all overloads completely is helpful for in-page search. 
* Usage notes below the set of overloads: This is fine, but the formatting seems to be missing blank lines between paragraphs. It looks kind of squished. See HTTPClient's first method.

What do you think?
* Overload titles: We discussed the idea of using the first sentence as the title for the overload. This would provide something more meaningful than "Overload 1", but it means we would need to be very deliberate in how we write those sentences. 

Thanks! This is coming along nicely.

petebacondarwin committed:
fix(docs-infra): update overload rendering

Based on the review here:
https://github.com/angular/angular/pull/24976#issuecomment-415535125

mary-poppins commented:
You can preview 8c80f61 at https://pr24976-8c80f61.ngbuilds.io/.

petebacondarwin commented:
Fix parameter table overflow

petebacondarwin committed:
fixup! fix(docs-infra): update overload rendering

mary-poppins commented:
You can preview 823a582 at https://pr24976-823a582.ngbuilds.io/.

jenniferfell reviewed (commented):
Looks good! Thanks for persevering with this.

gkalpak commented on aio/src/styles/2-modules/_api-pages.scss:
```diff
@@ -80,21 +205,78 @@
+
+    summary {
+      height: inherit;
+      padding: 0 0 8px;
```
I suggest removing the bottom padding (or at least keeping the top-bottom paddings symmetric.

Now:
![padding-before](https://user-images.githubusercontent.com/8604205/47502527-bcc89980-d870-11e8-9b2b-b3c0f185a37a.png)

Without padding:
![padding-after](https://user-images.githubusercontent.com/8604205/47502537-c18d4d80-d870-11e8-947e-07e0012bfa66.png)

gkalpak reviewed (changes requested):
Looks much better :100:

A couple more suggestion:

1. I find the `show`/`collapse` pair a little unexpected. I would expect either `show`/`hide` or `expand`/`collapse`. 

2. The overloads' `<summary>` element has the default `:focus` outline, which (a) doesn't look good and (b) is funnily deformed during the show/collapse transition:

    ![outline-before](https://user-images.githubusercontent.com/8604205/47502590-e681c080-d870-11e8-921b-00a36dfb8431.png)

    I suggest either removing the outline altogether or ensuring that the rotating carret icon does not affect it (e.g. with `... summary { overflow: hidden; }`):

    ![outline-after](https://user-images.githubusercontent.com/8604205/47502603-eaadde00-d870-11e8-8331-7624210aae12.png)

petebacondarwin commented on aio/src/styles/2-modules/_api-pages.scss:
```diff
@@ -1,3 +1,128 @@
+          vertical-align: middle;
+      }
+
+      tbody > tr > td tr td:first-child {
```
OK, so this is overriding this specific rule: https://github.com/angular/angular/blob/95993e1dd523cff96ebc2b42beb1e75dc95ca049/aio/src/styles/2-modules/_table.scss#L58-L62

```
    tr td:first-child {
      @media (max-width: 480px) {
        background-color: $lightgray;
      }
    }
```

petebacondarwin replied:
I don't want to start touching that file but perhaps @sjtrimble will fix this when she looks at our responsive table situation.

gkalpak replied:
Fair enough :grin:

petebacondarwin commented on aio/tools/transforms/templates/api/lib/memberHelpers.html:
```diff
@@ -140,15 +151,16 @@ <h3>
   {% if method.description -%}
     <tr>
       <td class="description">
```
This (short-)description is below the `.method-table` table, and the other (short-)description is below the `.overload-info` div, so they are distinguishable. I'd prefer not to have to start prefixing all these classes, otherwise you'd end up with `overload-info-short-description` etc.

gkalpak replied:
🤔

petebacondarwin commented on aio/tools/transforms/templates/api/lib/memberHelpers.html:
```diff
@@ -40,54 +40,55 @@
 {%- macro renderOverloadInfo(overload, cssClass, method) -%}
+<div class="overload-info">
+  {% if overload.shortDescription and (overload.shortDescription != method.shortDescription) %}
+  <div class="short-description">
```
see https://github.com/angular/angular/pull/24976#discussion_r228704439

petebacondarwin committed:
fixup! fix(docs-infra): update overload rendering

mary-poppins commented:
You can preview da6d2bf at https://pr24976-da6d2bf.ngbuilds.io/.

petebacondarwin commented:
Another review needed please @gkalpak (according to PullApprove).
